### PR TITLE
API-14: Support for FS style upload workflow

### PR DESCRIPTION
**Pull request by Arpit-Sharma-USC**
-logic to support new FS api based file upload
-update util::wrapped_request to support returning both json and standard responses for FS api
-update util::api_endpoint_url to support both FS-v7 and v4 endpoints url

[JIRA](https://bynder.atlassian.net/browse/API-14)

### CODE REVIEW

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]:
```suggestion
        return response["file_id"]
```

File: bynder_sdk/util.py
Comment (githuib) [inline bynder_sdk/util.py]:
```suggestion
        endpoint = api_endpoint_url(self, endpoint, is_fs_endpoint)
```

File: bynder_sdk/util.py
Comment (githuib) [inline bynder_sdk/util.py]: Feels somewhat cumbersome to have to pass `is_fs_endpoint` now everytime. And say in the future some new api gets added, we would need to add another arg. Maybe better to just check if it starts with `/v6/` or `/v7/`?

Or just prefixing the other routes with `/api` and always returning `'https://{}{}'` might be the most robust / future-proof solution.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: They changed the `finalise` into `finalise_api` now for api usage.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]:
```suggestion
            _, file_name = os.path.split(file_path)
```

File: bynder_sdk/util.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/util.py]: Sounds good!

File: bynder_sdk/util.py
Comment (githuib) [inline bynder_sdk/util.py]: Not all "old" routes start with `v4` though. The pybynder endpoints are just `/api/something`.

So maybe instead check if it starts with /v6 or /v7. Or just add "/api" in front of the old endpoints so we don't have to do this check here and it is more explicit to the reader of the code.

File: bynder_sdk/util.py
Comment (githuib) [inline bynder_sdk/util.py:22]: Instead of introducing this `need_response_json` arg, couldn't we just check somehow in the `response` object if it has a json body?

Or maybe something like this would do the trick?
```
try:
    return response.json()
except ValueError:
    return response
```
(and ditch parse_json_for_response() since we wouldn't need it anymore)

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]:
```suggestion
            for chunk_nr, chunk in enumerate(_read_in_chunks(f)):
                self._upload_chunk(file_id, chunk, chunk_nr)
```

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]:
```suggestion
        self.session.headers['content-sha256'] = sha256(data).hexdigest()
```
(assuming `hexdigest()` returns a string)

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]:
```suggestion
            '/v7/file_cmds/upload/{}/chunk/{}'.format(file_id, chunk_nr),
```
instead of the variable that's only used once

File: bynder_sdk/util.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/util.py:22]: Good idea 😄

File: bynder_sdk/client/upload_client.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py]: Nice!

**Review by betacar — DISMISSED** [review]

File: bynder_sdk/client/asset_bank_client.py
Comment (WouterToering) [inline bynder_sdk/client/asset_bank_client.py:8]: I'll comment here since it's a bit easier to have a "threaded" conversation this way:

Something that would be great to address with this PR is the fragility of the upload process regarding failed requests. With both the old and the new code uploads are basically ruined when a single request fails which is not very user friendly. If say for example chunk 50000/50001 fails, we should retry it rather than going "welp your upload failed, try again". Same principle goes for the finalise endpoint (and perhaps media save endpoint if a retry would make sense there).

Comment (WouterToering) [reply bynder_sdk/client/asset_bank_client.py:8]: As discussed we will pick this up in another ticket.

**Review by WouterToering — DISMISSED** [review]

File: bynder_sdk/client/asset_bank_client.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/asset_bank_client.py:8]: Alright, great observation I must say. This makes good sense to do to me.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: Where would the user need the `file_id` or `correlation_id` for? I'd say those are implementation details.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: Could be inlined, although I don't think we even need the correlation_id anyway?

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: It could also contain some other fields, such as the asset name: https://bynder.docs.apiary.io/#reference/upload-assets/6-saving-a-completed-upload/save-as-a-new-asset

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: It could also contain some other fields, such as the asset name: https://bynder.docs.apiary.io/#reference/upload-assets/6-saving-a-completed-upload/save-as-a-new-asset

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py:107]: Would make more sense maybe to split this up in two methods, since they're completely different?

And also good to check if the brandId is present in the data maybe, since it's a required arg in the body.

File: bynder_sdk/client/upload_client.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py]: My understanding was the user can get back to us with a correlation Id and the file id in case something goes wrong with the upload to help diagnose the issue. But I can get rid of it for consistency with the other sdks.

File: bynder_sdk/client/upload_client.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py]: I will remove the return altogether.

File: bynder_sdk/client/upload_client.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py]: Yeah let me make this more generic like `asset information`

File: bynder_sdk/client/upload_client.py
Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py:107]: I agree there can be a check for the brandId but they can be 2 methods or even 1 method since the functionality of the method is just to save the media and change the url based on a param personally I would prefer to have it as 1 method.

File: bynder_sdk/client/asset_bank_client.py
Comment (betacar) [inline bynder_sdk/client/asset_bank_client.py]: I believe `brand_id` is no longer required.

Comment (betacar) [reply bynder_sdk/client/asset_bank_client.py]: Now that I think of it, the whole `query` hash is no longer used.

Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/asset_bank_client.py]: No I think the `brand_id` is still needed. It's not needed when we update an existing asset (in which case we specify the `mediaId`) which is why we unset it under `save_media` when `mediaId` is present. 

But if it's a new asset upload we need to associate it with a brandId. See [here](https://bynder.docs.apiary.io/#reference/download/download-operations/save-as-a-new-asset) 

There are other optional form fields too like `name` and `description `for instance which will be part of the `query` aka upload_data.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [reply bynder_sdk/client/upload_client.py:107]: Yeah I thought two methods would make the code a bit simpler, since with one method it would basically be one if/else with the two requests that have different args, except for the file_id.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: brandId (or a body in general) is only required for the case where no media ID is passed. And I wonder why the `data['brandId'].strip()` check is there?

Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py]: the `strip()` check is there to make sure the client doesn't pass something like `'   '` to just escape the required `brandId`. So instead of hitting our endpoints and get a response that the `brandId` is invalid, we just do it here before even hitting the endpoint.

This is just a sanity check tbh.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [inline bynder_sdk/client/upload_client.py]: The intent can be left out from the SDK requests. It will fallback to the correct one server side.

File: bynder_sdk/client/upload_client.py
Comment (githuib) [reply bynder_sdk/client/upload_client.py]: Hmm sounds a bit arbitrary to me though why just for this field. Besides, in case it would have spaces in it, we would still pass that as part of the data. So if we want to do stuff like stripping spaces, I'd rather do that as part of some validating step in general (but I wouldn't recommend doing that in this PR and maybe it would be a bit overkill in general anyway).

Comment (Arpit-Sharma-USC) [reply bynder_sdk/client/upload_client.py]: Alright I will drop the `strip()` 👍

**Comment by Arpit-Sharma-USC** [comment]
As we all discussed, I created a branch `FS_upload_dev` to merge this work into.

I won't merge this to `FS_upload_dev` yet so it is easy for you to look at the changes and conversations we had on this PR. Once this is approved I will create a PR to merge `FS_upload_dev` into `master`. Which will resolve our goals for the week for `python-sdk`

**Review by githuib — APPROVED** [review]

**Pull request merged**
